src/Umbraco.Infrastructure/Persistence/Repositories/Implement/SimilarNodeName.cs

```diff
@@ -1,118 +1,249 @@
+                // filter items based on full name with suffix
+                items = items.Where(x => x.Text.InvariantStartsWith(model.FullName));
+                var secondarySuffix = GetFirstSuffix(items);
+                model.SetSuffix(secondarySuffix);
 
-                if (xpos < 0 || ypos < 0 || xpos != ypos)
-                    return string.Compare(xname, yname, StringComparison.Ordinal);
+                return model.FullName;
+            }
+
+            // has suffix - name without suffix also exists, therefore we simply increment
+            if (model.HasSuffix() && items.SimpleNameExists(model.Text))
+            {
+                var nextSuffix = GetSuffixNumber(items);
+                model.SetSuffix(nextSuffix);
 
-                // compare the part before (number)
-                var n = string.Compare(xname, 0, yname, 0, xpos, StringComparison.Ordinal);
-                if (n != 0)
-                    return n;
+                return model.FullName;
+            }
 
-                // compare (number) lengths
-                var diff = xname.Length - yname.Length;
-                if (diff != 0) return diff < 0 ? -1 : +1;
+            return name;
+        }
 
-                // actually compare (number)
-                var i = xpos;
-                while (i < xname.Length - 1)
+        private static int GetFirstSuffix(IEnumerable<StructuredName> items)
+        {
+            const int suffixStart = 1;
+
+            if (!items.Any(x => x.Number == suffixStart))
+            {
+                // none of the suffixes are the same as suffixStart, so we can use suffixStart!
+                return suffixStart;
+            }
+
+            return GetSuffixNumber(items);
+        }
+
+        private static int GetSuffixNumber(IEnumerable<StructuredName> items)
+        {
+            int current = 1;
+            foreach (var item in items.OrderBy(x => x.Number))
+            {
+                if (item.Number == current)
                 {
-                    if (xname[i] != yname[i])
-                        return xname[i] < yname[i] ? -1 : +1;
-                    i++;
+                    current++;
+                }
+                else if (item.Number > current)
+                {
+                    // do nothing - we found our number!
+                    // eg. when suffixes are 1 & 3, then this method is required to generate 2
+                    break;
                 }
-                return 0;
             }
+
+            return current;
         }
 
-        // gets a unique name
-        public static string GetUniqueName(IEnumerable<SimilarNodeName> names, int nodeId, string nodeName)
+        internal class StructuredName
         {
-            var uniqueNumber = 1;
-            var uniqueing = false;
-            foreach (var name in names.OrderBy(x => x, new Comparer()))
+            private const string Suffixed_Pattern = @"(.*)\s\((\d+)\)$";
```

Comment:
Not sure we should use "\s" here.  Why not force it to be only " " (Space)
```suggestion
            private const string Suffixed_Pattern = @"(.*) \((\d+)\)$";
```

Reply:
Agree... we can force a space character.